library/private/options_tools.h.in

```diff
@@ -0,0 +1,340 @@
+    i++;
+  }
+  return stream.str();
+}
+
+//----------------------------------------------------------------------------
+/**
+ * Generic templated string generation from provided value
+ * rely on std::to_string
+ */
+template<typename T>
+static std::string toString(const T& var)
+{
+  return std::to_string(var);
+}
+
+//----------------------------------------------------------------------------
+/**
+ * Generic templated string generation from provided value with an option_types
+ * rely on options_tools::toString and ignore option_types
+ */
+template<typename T>
+static std::string toString(const T& value, option_types)
+{
+  return options_tools::toString(value);
+}
+
+//----------------------------------------------------------------------------
+/**
+ * Specific string generation from provided double value with an option_types
+ * rely on options_tools::toString and uses option_types to distinguish between
+ * double and ratio_t
+ */
+static std::string toString(const double& value, option_types type)
+{
+  if (type == option_types::_ratio)
+  {
+    ratio_t ratio = value;
+    return options_tools::toString(ratio);
+  }
+  else
+  {
+    return options_tools::toString(value);
+  }
+}
+
+//----------------------------------------------------------------------------
+/**
+ * Generated method that set an option in provided options according to
+ * a provided name and value.
+ * Can throw options::inexistent_exception or options::incompatible_exception
+ */
+void set(options& opt, const std::string& name, const option_variant_t& value)
+{
+  try
+  {
+    ${_options_setter};
+    else throw options::inexistent_exception("Option " + name + " does not exist");
+  }
+  catch (const std::bad_variant_access&)
+  {
+    throw options::incompatible_exception("Trying to set " + name + " with incompatible type");
+  }
+}
+
+//----------------------------------------------------------------------------
+/**
+ * Generated method that get an option value in
+ * provided options using a provided name
+ * Can throw options::inexistent_exception
+ */
+option_variant_t get(const options& opt, const std::string& name)
+{
+  option_variant_t var;
+  ${_options_getter};
+  else throw options::inexistent_exception("Option " + name + " does not exist");
+  return var;
+}
+
+//----------------------------------------------------------------------------
+/**
+ * Generated method that get an option type using a provided name
+ * Can throw options::inexistent_exception
+ */
+option_types getType(const std::string& name)
+{
+  ${_options_typer};
+  else throw options::inexistent_exception("Option " + name + " does not exist");
+  // Cannot be reached
+  return option_types::_bool;
+}
+
+//----------------------------------------------------------------------------
+/**
+ * Generated method to recover all option names
+ */
+std::vector<std::string> getNames()
+{
+  std::vector<std::string> vec{ ${_options_lister} };
+  return vec;
+}
```

Comment:
This would be definitely more performant (optimized away during compilation), but it will require more changes I suppose so feel free to ignore if it's too complicated
```suggestion
constexpr std::array<std::string_view> getNames()
{
  std::array<std::string_view> vec{ ${_options_lister} };
  return vec;
}
```

Reply:
Ill add an item to improve compile-time opti for this file.

First bump is that I need to know the number of options at compile time, which require more work in the json parsing stuff.